### PR TITLE
Handle multiline multibyte headers correctly

### LINKED ISSUE #194: HeaderWrap::mimeDecodeValue unable to handle multiline utf header splitted across character
I've encountered email with subject, that was not handled by *zend-mail* properly. After some investigation I found that header value was split in middle of **multibyte** utf character. 

Test to reproduce it:

```php
    public function testMultilineWithMultibyteSplitAcrossCharacter()
    {
        $originalValue = 'аф';

        $this->assertEquals(strlen($originalValue), 4);

        $part1 = base64_encode(substr($originalValue, 0, 3));
        $part2 = base64_encode(substr($originalValue, 3));

        $header = 'Subject: =?utf-8?B?' . $part1 . '?==?utf-8?B?' . $part2 . '?=';

        $this->assertEquals(
            $originalValue,
            GenericHeader::fromString($header)->getFieldValue()
        );
    }
```
I fix it locally:

```php
    public static function mimeDecodeValue($value)
    {
        return array_reduce(
            imap_mime_header_decode(imap_utf8($value)),
            function ($acc, $item) { return $acc . $item->text; },
            ''
        );
    }
```

If depending on *imap* extension is ok, I will make PR. Or we can check if *imap* extension is enabled, otherwise use `iconv_mime_decode` as before.

### PR DESCRIPTION
**Pull request by fred84**
Fixes #194

### CODE REVIEW

File: test/Header/HeaderWrapTest.php [hunk 1]
```diff
@@ -111,4 +111,25 @@ public function testCanBeEncoded()
         $res = HeaderWrap::canBeEncoded($value);
         $this->assertTrue($res);
     }
+
+    public function testMultilineWithMultibyteSplitAcrossCharacter()
+    {
+        if (! extension_loaded('imap')) {
+            $this->markTestSkipped('works only with imap extension enabled');
+        }
```
Comment (glensc) [inline test/Header/HeaderWrapTest.php]: perhaps phpunit annotation instead?

```
@requires extension imap
```

Comment (fred84) [reply test/Header/HeaderWrapTest.php]: @glensc Thanks for suggestion! PR updated.

File: src/Header/HeaderWrap.php [hunk 2]
```diff
@@ -112,9 +112,27 @@ public static function mimeDecodeValue($value)
 
         $decodedValue = iconv_mime_decode($value, ICONV_MIME_DECODE_CONTINUE_ON_ERROR, 'UTF-8');
 
+        // imap (unlike iconv) can handle multibyte headers which are splitted across multiple line
+        if (self::isNotDecoded($value, $decodedValue) && extension_loaded('imap')) {
+            return array_reduce(
+                imap_mime_header_decode(imap_utf8($value)),
+                function ($acc, $item) {
```
Comment (glensc) [inline src/Header/HeaderWrap.php]: perhaps more useful variable names than `$acc` and `$item`?

Comment (fred84) [reply src/Header/HeaderWrap.php]: updated

File: src/Header/HeaderWrap.php [hunk 3]
```diff
@@ -112,9 +112,27 @@ public static function mimeDecodeValue($value)
...
+                ''
+            );
+        }
+
         return $decodedValue;
     }
 
+    private static function isNotDecoded($originalValue, $value)
+    {
+        return substr($value, 0, 2) == '=?'
```
Comment (samsonasik) [inline src/Header/HeaderWrap.php]: can use `0 === strpos($value, '=?')`

Comment (fred84) [reply src/Header/HeaderWrap.php]: done

**Comment by fred84** [comment]
@samsonasik @glensc Some tests on DEPS=lowest are broken, but it's unrelated to my changes. Master branch also broken.

**Comment by weierophinney** [comment]
Thanks, @fred84!

**Pull request merged**